runtime/wasm/src/host_exports.rs

```diff
@@ -335,6 +343,67 @@ where
         )
     }
 
+    // Read the IPFS file `link`, split it into JSON objects, and invoke
+    // the exported function `callback` on each JSON object. The successful
+    // return value contains all entity operations that were produced by the
+    // callback invocations. Each invocation of `callback` happens in its own
+    // instance of a WASM module, which is identical to `module` when it was
+    // first started.
+    pub(crate) fn ipfs_map(
+        &self,
+        module: &WasmiModule<E, L, S, U>,
+        link: String,
+        callback: &str,
+        flags: Vec<String>,
+    ) -> Result<Vec<EntityOperation>, HostExportError<impl ExportError>> {
+        const JSON_FLAG: &str = "json";
+        if !flags.contains(&JSON_FLAG.to_string()) {
+            return Err(HostExportError(format!("Flags must contain 'json'")));
+        }
+
+        let valid_module = module.valid_module.clone();
+        let ctx = module.ctx.clone();
+        let callback = callback.to_owned();
+        // Create a base error message to avoid borrowing headaches
+        let errmsg = format!(
+            "ipfs_map: callback '{}' failed when processing file '{}'",
+            &*callback, &link
+        );
+
+        let start = Instant::now();
+        let mut last_log = Instant::now();
+        let logger = ctx.logger.new(o!("ipfs_map" => link.clone()));
+        let operations = self.block_on(
+            self.link_resolver
+                .json_stream(&Link { link })
+                .and_then(move |stream| {
+                    stream
+                        .and_then(move |sv| {
+                            let module = WasmiModule::from_valid_module_with_ctx(
+                                valid_module.clone(),
+                                ctx.clone(),
+                            )?;
+                            let result = module.handle_json_callback(&*callback, &sv.value);
+                            // Log progress every 15s
+                            if last_log.elapsed() > Duration::from_secs(15) {
+                                debug!(
+                                    logger,
+                                    "processed {} lines in {}s so far",
```

Comment:
Our log messages start with a capital first letter unless it's a symbol like `ipfs_map`.

Reply:
Fixed